animation_nodes/algorithms/mesh/triangulate_mesh.pyx

```diff
@@ -0,0 +1,316 @@
+            index += 3
+
+    return newPolygons
+
+@cython.cdivision(True)
+def triangulatePolygonsUsingEarClipMethod(Vector3DList vertices, PolygonIndicesList polygons):
+    cdef unsigned int *oldPolyLengths = polygons.polyLengths.data
+    cdef Py_ssize_t polygonAmount = polygons.getLength()
+    cdef Py_ssize_t i, triAmount, polyLength
+
+    # Total number of triangle polygons.
+    triAmount = 0
+    for i in range(polygonAmount):
+        polyLength = oldPolyLengths[i]
+        triAmount += polyLength - 2
+
+    cdef unsigned int *oldIndices = polygons.indices.data
+    cdef unsigned int *oldPolyStarts = polygons.polyStarts.data
+
+    cdef PolygonIndicesList newPolygons = PolygonIndicesList(
+                                          indicesAmount = 3 * triAmount,
+                                          polygonAmount = triAmount)
+    cdef unsigned int *newIndices = newPolygons.indices.data
+    cdef unsigned int *newPolyStarts = newPolygons.polyStarts.data
+    cdef unsigned int *newPolyLengths = newPolygons.polyLengths.data
+
+    cdef float angle
+    cdef Vector3 v1, v2, v3
+    cdef Vertex* verticesData
+    cdef Py_ssize_t polyStart, triangleIndex, polyIndex
+    cdef Py_ssize_t j, index, previousIndex, earIndex, nextIndex
+
+    polyIndex = 0
+    triangleIndex = 0
+    for i in range(polygonAmount):
+        polyStart = oldPolyStarts[i]
+        polyLength = oldPolyLengths[i]
+
+        # Initialization of polygon's vertices.
+        verticesData = <Vertex*>PyMem_Malloc(polyLength * sizeof(Vertex))
+        projectPolygonVertices(vertices, polyStart, polyLength, oldIndices, verticesData)
+
+        # Check polygon polarity.
+        if not polyCounterClockwise(verticesData):
+            for j in range(polyLength):
+                previousIndex = verticesData[j].previous
+                nextIndex = verticesData[j].next
+                verticesData[j].previous = nextIndex
+                verticesData[j].next = previousIndex
+
+        # Calculate angle and status of vertices.
+        for j in range(polyLength):
+            angle = calculateAngleWithSign(verticesData, j)
+            verticesData[j].angle = angle
+            if angle >= 0.0 and not pointsInTriangle(polyLength, verticesData, j):
+                verticesData[j].isEar = True
+            else:
+                verticesData[j].isEar = False
+
+        # Calculate triangle polygon indices.
+        for j in range(polyLength - 2):
+            if polyLength - 2 - j == 1:
+                newPolyStarts[triangleIndex] = polyIndex
+                newPolyLengths[triangleIndex] = 3
+                triangleIndex += 1
+
+                index = 0
+                for k in range(polyLength):
+                    if verticesData[k].isEar:
+                        newIndices[polyIndex + index] = verticesData[k].index
+                        index += 1
+                        if index == 3: break
+                polyIndex += 3
+                break
```

Comment:
It appears you are checking if this is the last triangle and skipping the extra checks if it is. What I would do is move this check after the loop. So you loop one time less `for j in range(polyLength - 3):` and then add the last triangle manually.

Reply:
It is done.